src/signet.cpp

```diff
@@ -0,0 +1,150 @@
+
+static constexpr uint8_t SIGNET_HEADER[4] = {0xec, 0xc7, 0xda, 0xa2};
+
+static constexpr unsigned int BLOCK_SCRIPT_VERIFY_FLAGS = SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_DERSIG | SCRIPT_VERIFY_NULLDUMMY;
+
+static bool FetchAndClearCommitmentSection(const Span<const uint8_t> header, CScript& witness_commitment, std::vector<uint8_t>& result)
+{
+    CScript replacement;
+    bool found_header = false;
+    result.clear();
+
+    opcodetype opcode;
+    CScript::const_iterator pc = witness_commitment.begin();
+    std::vector<uint8_t> pushdata;
+    while (witness_commitment.GetOp(pc, opcode, pushdata)) {
+        if (pushdata.size() > 0) {
+            if (!found_header && pushdata.size() > (size_t) header.size() && Span<const uint8_t>(pushdata.data(), header.size()) == header) {
+                // pushdata only counts if it has the header _and_ some data
+                result.insert(result.end(), pushdata.begin() + header.size(), pushdata.end());
+                pushdata.erase(pushdata.begin() + header.size(), pushdata.end());
+                found_header = true;
+            }
+            replacement << pushdata;
+        } else {
+            replacement << opcode;
+        }
+    }
+
+    if (found_header) witness_commitment = replacement;
+    return found_header;
+}
+
+static uint256 ComputeModifiedMerkleRoot(const CMutableTransaction& cb, const CBlock& block)
+{
+    std::vector<uint256> leaves;
+    leaves.resize(block.vtx.size());
+    leaves[0] = cb.GetHash();
+    for (size_t s = 1; s < block.vtx.size(); ++s) {
+        leaves[s] = block.vtx[s]->GetHash();
+    }
+    return ComputeMerkleRoot(std::move(leaves));
+}
+
+SignetTxs SignetTxs::Create(const CBlock& block, const CScript& challenge)
+{
+    CMutableTransaction tx_to_spend;
+    tx_to_spend.nVersion = 0;
+    tx_to_spend.nLockTime = 0;
+    tx_to_spend.vin.emplace_back(COutPoint(), CScript(OP_0), 0);
+    tx_to_spend.vout.emplace_back(0, challenge);
+
+    CMutableTransaction tx_spending;
+    tx_spending.nVersion = 0;
+    tx_spending.nLockTime = 0;
+    tx_spending.vin.emplace_back(COutPoint(), CScript(), 0);
+    tx_spending.vout.emplace_back(0, CScript(OP_RETURN));
+
+    // can't fill any other fields before extracting signet
+    // responses from block coinbase tx
+
+    // find and delete signet signature
+    if (block.vtx.empty()) return {tx_to_spend, tx_spending}; // no coinbase tx in block; invalid
+    CMutableTransaction modified_cb(*block.vtx.at(0));
+
+    const int cidx = GetWitnessCommitmentIndex(block);
+    if (cidx == NO_WITNESS_COMMITMENT) {
+        return {tx_to_spend, tx_spending}; // require a witness commitment
+    }
+
+    CScript& witness_commitment = modified_cb.vout.at(cidx).scriptPubKey;
+
+    std::vector<uint8_t> signet_solution;
+    if (!FetchAndClearCommitmentSection(SIGNET_HEADER, witness_commitment, signet_solution)) {
+        // no signet solution -- allow this to support OP_TRUE as trivial block challenge
```

Comment:
It does enable `generate*` for `OP_TRUE` signets, though, which is handy.

Reply:
Changing consensus code to return early to avoid having to write tests for it seems backward. I'd rather have no tests than a test that requires changing consensus code. Also, this will need to be changed in the BIP first.

Reply:
Honestly, I always assumed that an OP_TRUE challenge with an empty solution was implicitly accepted, even with the BIP as it stands now.

I read this code as "if it is able to fetch the signet commitment, deserialize it into the scriptSig/witness of the spending tx vin[0]", i.e. it doesn't actually return early, it just doesn't deserialize the above.

We can remove this an the OP_TRUE test portion for now, though. Thoughts, @ajtowns ?

Reply:
I think having "no block solution" encoded as no signet commitment rather than an empty signet commitment seems slightly better; and it makes it possible for signet with "pushdata(non-zero)" as the challenge act like a unique (as far as magic is concerned), pow-enforcing regtest without needing any additional mining code. I'm not sure that's useful, but I don't think it's worse -- you can have an empty challenge to require a signet commitment (with scriptSig "op_true") if you want. I think the BIP should be changed either way to clarify what happens if the signet commitment isn't present.

Also the BIP says the scriptPubKey is called the "block challenge", but the command line args is "-signet_blockscript"; seems inconsistent?

Reply:
I like challenge/solution, but `-signet_blockchallenge` looks really long to me...

Reply:
The BIP refers to it as `signet_challenge`, not block challenge?! Am I reading the wrong doc?

Reply:
BIP revision proposal: https://github.com/bitcoin/bips/pull/983

Reply:
Ahh, yeah it does call it `signet_challenge` in the new version.